client/network/src/service.rs

```diff
@@ -95,6 +96,14 @@ pub struct NetworkService<B: BlockT + 'static, H: ExHashT> {
 	peerset: PeersetHandle,
 	/// Channel that sends messages to the actual worker.
 	to_worker: TracingUnboundedSender<ServiceToWorkerMsg<B, H>>,
+	/// For each peer and protocol combination, an object that allows sending notifications to
+	/// that peer. Updated by the [`NetworkWorker`].
+	peers_notifications_sinks: Arc<Mutex<HashMap<(PeerId, ConsensusEngineId), NotificationsSink>>>,
```

Comment:
There's also a design issue here, which is that one `ConsensusEngineId` might map to multiple protocol names in the future (cc #6605). It's a bit unclear to met what to do yet.

Reply:
Permitting negotiation of alternative protocols, like different versions, is definitely desirable. Just having a concept of a "protocol group / family / list" within which protocols are treated as alternatives for negotiation may be enough, i.e. the ability to put multiple protocol names together in a list. The `ConsensusEngineId` should not be needed for that. In any case, I just mentioned it because in the comment just below for the field that maps `ConsensusEngineId` to protocol names it is referred to as legacy, yet here it is used as part of a key in a _new_ map.